Add initial draft of Auth GEP 1494

youngnick opened the pull request:
/kind gep

**What this PR does / why we need it**:

This is to get the conversation started around Auth* in Gateway API. Hopefully this part should not be too contentious yet, but reviews gratefully accepted.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Updates #1494

LiorLieberman commented:
/cc

LiorLieberman commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,131 @@
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTPRoutes.
+
+* Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected.
```
Does this imply that Ana has the ability to override the cluster-admin configuration? I assume some use cases for this functionality might involve testing, but I’m curious to hear about other potential use cases. My concern is that an app developer, who may not have a strong understanding of authentication, would have the ability to override cluster-admin (or security admin) defaults.

jgao1025 replied:
> Does this imply that Ana has the ability to override the cluster-admin configuration? I assume some use cases for this functionality might involve testing, but I’m curious to hear about other potential use cases. My concern is that an app developer, who may not have a strong understanding of authentication, would have the ability to override cluster-admin (or security admin) defaults.

I think some public pages, like a status page or help page, might not need authn? 

It could be more flexible if the Gateway API allowed an HTTPRoute to either override or inherit settings from the Gateway under a hierarchical control model. This flexibility would be especially useful in cases where roles overlap, such as in smaller organizations where Ana and Chihiro might be the same person.

youngnick replied:
Yes, the most likely use case here is for healthchecks, or being able to say "the whole website needs auth, except for `/public`" or something.

pubudu538 replied:
From my experience, cluster admins often lack a clear understanding of the deployed applications and their paths unless developers explicitly provide this information. In this context, the role of the cluster admin would be to enable default authentication globally for all routes. Developers, like Ana, can then override this default authentication at the httproute level or for specific routes within the httproute. WDYT?

youngnick replied:
Yes, this is roughly what I'm thinking here. A default, not an override, although I'm not sure we'll be using Policy Attachment to do this or not yet.

youngnick replied:
I updated this wording.

LiorLieberman commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,131 @@
+
+## Conformance Details
+
+(This section describes the names to be used for the feature or
+features in conformance tests and profiles.
+
+These should be `CamelCase` names that specify the feature as
+precisely as possible, and are particularly important for
+Extended features, since they may be surfaced to users.)
```
```suggestion
(from https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162/index.md#standardize-features-and-conformance-tests-names)

#### Feature Names

Every feature should:

1. Start with the resource name. i.e HTTPRouteXXX
2. Follow the PascalCase convention. Note that the resource name in the string should come as is and not be converted to PascalCase, i.e HTTPRoutePortRedirect and not HttpRoutePortRedirect.
3. Not exceed 128 characters.
4. Contain only letters and numbers

### Conformance tests 

Conformance tests file names should try to follow the the `pascal-case-name.go` format.
For example for `HTTPRoutePortRedirect` - the test file would be `httproute-port-redirect.go`.

Treat this guidance as "best effort" because we might have test files that check the combination of several features and can't follow the same format.

In any case, the conformance tests file names should be meaningful and easy to understand.
```

LiorLieberman replied:
created https://github.com/kubernetes-sigs/gateway-api/pull/3502 as well to fix the template

LiorLieberman reviewed (commented):
Thanks for very detailed content. Left a few comments

jgao1025 commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,131 @@
+
+## TLDR
+
+Provide a method for configuring **Gateway API implementations** to add HTTP Authentication for north-south traffic. The method may also include Authorization config if practical.
```
Hmm.. Are these auth* applicable only to inbound traffic, or do they cover both ingress and egress?

youngnick replied:
Theoretically, it could cover both, depending on how you orient the Gateway, but I've added something to say this is only ingress for now. Good point, thanks!

Ongy commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+TLS includes the possibility of having both the client and server present certificates for the other party to validate. (This is often called "mutual TLS", but is distinct from the use of that term in Service Mesh contexts, where it means something more like "mutual TLS with short-lifetime, automatically created and managed dynamic keypairs for both client and server").
+
+In this case, the server authenticates the client based on the client presenting a certificate that's signed by an authority that's also trusted by the server's trust chain. Some implementations also allow details about the certificate to be passed through to backend clients, to be used in authorization decisions.
```
I think the "also" is misplaced, since it's about the server also verifying a client, not about the client's cert being trusted by the same CA as ??.
The "trust chain" seems weird as well. Generally we have trust roots on a server, and the chain is presented by the client.
Since the client side authentication process of the server is implied prior knowledge, I think this can be condensed a bit.

I suggest:

`In this case, the server also authenticates the client, based on the certificate chain presented by the client. Some implementations also allow details about the certificate to be passed through to backend clients, to be used in authorization decisions`

youngnick replied:
Okay, thanks.

Ongy commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+In Basic HTTP Auth, a server asks for authentication as part of returning a `401` status response, and the client includes an `Authorization` header that includes a Base64-encoded username and password.
+
+Because the password is only _encoded_ and not _encrypted_, Basic Auth is totally unsafe when used outside of an encrypted session (like a HTTPS connection).
```
While the "raw" passwords in basic auth have additional issues (long lived, impersonating human on login pages), JWT and afaik OAUTH/OIDC are not generally not secure against replay attacks.

I.e. besides mTLS (which enforces confidentiality) all mechanisms are insecure in plaintext messages.

I think it's best to move the notion of requiring encryption for safe usage out of Basic Auth and potentially add a note here that basic auth has an additional issue with long-lived (potentially higher power) tokens being exchanged.

youngnick replied:
I agree with your points here, but the intent here was to say "nobody in their right mind uses Basic Auth outside of _at least_ an encrypted connection" (in a specifcation kind of way of course).

I'm not trying to document _all_ of the possible issues with every one of these, just to give a very basic overview so that everyone reading this who might not have all the context can get _some_ background.

I updated the wording here, PTAL.

Ongy replied:
I do get why it's here. The suggestion is mostly there to avoid someone taking the omission on other methods as sign that they are ok in an unencrypted context, when they should still be avoided.

OTOH, this isn't a tutorial about auth. And likely not front and center and the chances to trip up someone are low enough.

youngnick replied:
It's a fair point, but I'd like to keep it in mind and get the current state merged so we can start working on the API design.

Ongy commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+## Auth* User Stories
+
+
+* As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.
```
This might leak into the "API" phase, but I think there's 2 levels to this, which are both worth an explicit mention

* The API proposed implementation provides enough flexibility to integrate with an authorization mechanism and protect resources entirely in the gateway
* The API allows to inject information about the authentication result into the requests and allows bakend application to make authorization decisions based on this.

youngnick replied:
I agree with this point, and also that we're probably better to address this in the API phase. I'm going to leave this for now, and we'll come back to it in the next PR.

shaneutt replied:
(could compile comments like this that need to be accounted for later, but not as a part of this PR, as a TODO list right within the document)

youngnick replied:
(The API that we have in mind so far - basically promoting Envoy's ext_auth API to be a Gateway API one) does have the facility to handle both of these use cases. But we can't really start talking about until this PR merges.)

EternalDeiwos replied:
Regarding:

> ... ideally to be able to make Authorization decisions ...
   
> ... allows backend application to make authorization decisions based on this.
   
Wouldn't authorization (i.e. granting users' privileges, or otherwise producing credentials) normally be handled by an identity provider? At least in all except the basic auth case.

mikemorris replied:
> Wouldn't authorization (i.e. granting users' privileges, or otherwise producing credentials) normally be handled by an identity provider?

I think it depends - in OAuth/JWT cases yea that may be common, for other scenarios like mesh AuthZ policies, the infrastructure/identity provider just handles distribution of the authentication credentials (certificates used for mTLS in many implementations), and the actual policies are typically written as CRDs in the cluster by an platform engineer or developer, and enforced in the dataplane near the destination app, typically by a proxy.

(But this is getting _way_ ahead of the scope we want to immediately tackle to start.)

Ongy commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+* As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.
+* As Chihiro the Cluster Admin, I wish to be able to configure default Authentication settings (at least), with an option to enforce Authentication settings (preferable but not required) for some set of services exposed via Gateway API inside my cluster.
+* More User Stories welcomed here!
```
As Ana the Application Developer, I wish to be able to redirect users to a login page when they lack authentication, while unauthenticated API access gets the proper 40x response.

^ to make it clear that we need to handle "human" clients (browsers) slightly different to API consumers due to 30X/40X conventions.

Ongy reviewed (commented):
Couple comments around the auth mechanisms + suggestions for user stories.

New to k8s upstream work. Do you care about style comments? E.g. defining AuthN/AuthZ twice, or would you prefer to keep this to technical bits?

JackMyers001 commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTPRoutes.
+
+* Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected.
```
As I understand it, the current proposal *only* allows for "add auth to this website, except for this route". Could this be expanded to have some kind of rule-based authentication selection/bypass mechanism? I'd love to have the ability to choose a specific authentication mechanism (or bypass auth entirely) on a per-route basis, based on multiple factors (e.g. client IP, user agent).

This would make the Gateway API an amazing Layer 7 firewall, but I'm not sure if the project wants to support these kinds of capabilities; I saw [this proposal](https://github.com/kubernetes-sigs/gateway-api/blob/99a3934c6bc1ce0874f3a4c5f20cafd8977ffcb4/geps/gep-735/index.md?plain=1#L12) was closed partly because the feature "operated too much like a WAF/firewall".

Proposed user stories:

```md
* As Ana the Application Developer, I'm maintaining a legacy application that doesn't have any existing authentication mechanisms. I wish to enforce SSO when a user accesses the application through a browser. However, I wish to bypass authentication when the user accesses the application through the mobile app (based on the user agent), as it can't handle the SSO flow.
* As Ana the Application Developer, I wish to be able to use different authentication mechanisms for the same route based on the context of the request. I wish to use SSO when a user accesses my API through a browser, and use JWT when using my mobile app,
* As Chihiro the Cluster Admin, I wish to block undesirable User Agents or IP ranges (spiders, scrapers etc.) from accessing any services exposed by Gateway API inside my cluster.
* As Chihiro the Cluster Admin, I wish to be able to bypass AuthN when requests come in from a trusted, controlled IP range, so external services can access APIs running on my cluster without issue
```

youngnick replied:
I agree that all of those are interesting use cases, but I think that we need to handle the simpler (enable/disable) case first.

robscott commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTPRoutes.
```
Since we're saying this is exclusively for north-south traffic, I think I'd rather attach this config at the Gateway level. Maybe it would be safer to say something like this:
```suggestion
* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.
```

robscott commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,144 @@
+
+Basic auth is defined in [RFC-7617](https://datatracker.ietf.org/doc/html/rfc7617).
+
+#### TLS Client Certificate Authentication
```
May be worth referring to https://gateway-api.sigs.k8s.io/geps/gep-91/ as this is already in progress.

youngnick commented:
I've just pushed an update that addresses most comments, but feel like I should clarify again: This is by no means the final state here. This is intended just to get us started, so that we can _roughly_ agree on the goals and terminology used. We're not aiming for final quality for these sections, I expect them to get some updates as we work on the API.

shaneutt commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,151 @@
+* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.
+
+* Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected. This would probably need to work something like a default enabling at Gateway level, that can be specifically set at lower levels, but further design is TBD.
```
I think it might be helpful to draw out some motivation which explains this particular goal. Then, if we're going to provide a mechanism for the developer to override something [that is security related] set by the cluster administrator, we may want to consider flags to enable this override that Chihiro has to explicitly turn on.

mikemorris replied:
This feels less like flags and more like the defaults/overrides of the policy attachment model.

Some example HTTP endpoints that may be opted out of a default AuthN policy could be `/login` (for flows other than client-cert mTLS, unauthenticated human users will often need to get redirected _somewhere_ to authenticate and acquire a credential), or `/health` (some service health-checking mechanisms may prefer to avoid the overhead of authentication for performance reasons).

shaneutt replied:
Interesting :thinking:  An explanation of the motivation seems helpful in any case and maybe some elaboration here on the goal since apparently we two had somewhat different interpretations of it.

youngnick replied:
I've updated with some other wording, PTAL. In the interest of time, we could also move this to Deferred Goals, and get the basic setup handled first (with an eye to being able to handle something like this in the future).

shaneutt commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,151 @@
+
+#### TLS Client Certificate Authentication
+
+TLS includes the possibility of having both the client and server present certificates for the other party to validate. (This is often called "mutual TLS", but is distinct from the use of that term in Service Mesh contexts, where it means something more like "mutual TLS with short-lifetime, automatically created and managed dynamic keypairs for both client and server").
```
:+1:

shaneutt reviewed (commented):
Great start, and I appreciate you taking the time to try and get this ball moving again :saluting_face: 

/cc @kflynn @howardjohn @mikemorris

mikemorris commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,151 @@
+|[ingress-nginx](https://kubernetes.github.io/ingress-nginx/examples/customization/external-auth-headers/)|[httpbin](https://httpbin.org) ([Basic](https://kubernetes.github.io/ingress-nginx/examples/auth/external-auth/), [OAuth](https://kubernetes.github.io/ingress-nginx/examples/auth/oauth-external-auth/))|[Basic](https://kubernetes.github.io/ingress-nginx/examples/auth/basic/), [Client Certific ...
+|[Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto)|[External Authorization server (ext_authz filter)](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto) |[JWT](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto)|
+|[Contour](https://projectcontour.io/docs/1.24/config/client-authorization/)|[Envoy](https://projectcontour.io/docs/1.24/config/client-authorization/)|[Envoy(JWT)](https://projectcontour.io/docs/1.24/config/jwt-verification/)|
+|[Istio](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/)|[Envoy + istio (peer authN)](https://istio.io/latest/docs/concepts/security/#mutual-tls-authentication)|[JWT (Request authN)](https://istio.io/latest/docs/concepts/security/#authentication)|
```
```suggestion
|[Istio](https://istio.io/latest/docs/tasks/security/authorization/)|[mutual TLS ingress gateway](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway), [External Authorization](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/)|[JWT (RequestAuthentication)](https://istio.io/latest/docs/reference/config/security/request_authentication/)|
```
I think the mutual TLS ingress gateway docs are a more appropriate link for external auth than PeerAuthentication, which is focused on service-to-service configuration and out of scope for this GEP. Moved the External Authorization link to be more explicit, that feature is more specific than general external-facing (ingress) authentication. Updated link for JWT RequestAuthentication to the full doc page.

I'm unsure if it may also be worth referencing the [JWT claim-based routing](https://istio.io/latest/docs/tasks/security/authentication/jwt-route/) capabilities, which AFAIK are only configurable via an Istio VirtualService, or if any more specific documentation is worth linking to regarding External Authorization for an ingress gateway or for services exposed via an ingress gateway.

@howardjohn @keithmattix if I missed any documented functionality that may be relevant

youngnick replied:
Updated, thanks.

mikemorris commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,151 @@
+
+## Currently implemented Auth mechanisms in implementations
+
+Many Gateway API implementations have implemented some method of configuring Auth* that are currently Implementation Specific.
```
At least for the Istio examples, these aren't actually using Gateway API implementation-specific extensibility mechanisms (HTTPRoute filters, etc), but rather custom annotations and CRDs. It may be worth clarifying this statement a bit to better reflect if these are intended to be Gateway API well-known or ad-hoc extension examples, or just capabilities of implementations which also support Gateway API configuration.

mikemorris reviewed (commented):
Left a few suggestions for clarification, overall looks good to me for a provisional GEP, excited to make some progress on this!

youngnick commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,151 @@
+
+* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.
```
Given how much discussion we've generated with the current scope, I've been very reluctant to consider adding any more.

I'll put into "deferred goals" thinking about how this may be extended for other protocols.

shaneutt reviewed (approved):
Good first big step.

/approve

robscott commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,158 @@
+
+* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.
+
+* Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected. This would probably need to work something like a default enabling at Gateway level, that can be specifically set at lower levels, but further design is TBD. This goal comes from the relatively-common desire for Chihiro to be  ...
```
Or reading further, maybe this is a "deferred goal"

youngnick replied:
I think Stretch Goal is better, because it would be easy to accidentally prohibit this unless we think about it a bit at API design time.

robscott commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,158 @@
+## Deferred Goals
+
+* (Not performed during the Provisional Phase) Defining the API or configuration to be used.
+* Handling GRPCRoute (We will handle plain HTTP first)
```
Slight change to avoid any suggestion on whether this should be attached at a Route or Gateway level.
```suggestion
* Handling GRPC (We will handle plain HTTP first)
```

robscott reviewed (approved):
Thanks for writing this up @youngnick!

/approve

youngnick commented:
/retest

youngnick commented:
/retest

mlavacca reviewed (approved):
Thanks for putting this together!

/lgtm

youngnick commented:
/hold cancel

kflynn commented:
I got far more swamped than I'd wanted to be, but even though this is closed: I think that this is a great definition of common ground for digging into auth, nicely done @youngnick and @jgao1025! 

I do have one very minor comment beyond that: I'd vote for using just `Auth` instead of `Auth*` -- the asterisk keeps making me look for a footnote. 😂 

Again, nice work!

karthikbox commented on geps/gep-1494/index.md:
```diff
@@ -0,0 +1,160 @@
+* As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.
+* As Ana the Application Developer, I wish to be able to redirect users to a login page when they lack authentication, while unauthenticated API access gets the proper 40x response.
+* As Chihiro the Cluster Admin, I wish to be able to configure default Authentication settings (at least), with an option to enforce Authentication settings (preferable but not required) for some set of services exposed via Gateway API inside my cluster.
+* More User Stories welcomed here!
```
To add a a few (feel free chose or ignore):

1. As Ana the Application Developer, I want to integrate my Gateway API with an external authorization service like Okta or Auth0, so that I can leverage existing enterprise identity providers without implementing custom authentication logic.

3. As Ana the Application Developer, I want to validate JWTs against a configurable JWKS endpoint so that I can use tokens issued by different identity providers across different environments (development, staging, production).

4. As Ana the Application Developer, I need to simultaneously support multiple authentication methods with clear precedence rules (like mTLS for service clients and JWT for user requests), ensuring each client type can authenticate appropriately while maintaining consistent identity determination when multiple credentials are present.

I got more ideas around monitoring failures, caching JWT validation results, fallback, but i'll save those for later.
